src/Development/IDE/Plugin/CodeAction.hs

```diff
@@ -1108,17 +1109,31 @@ rangesForBinding' _ _ = []
 --       import (qualified) A (..) ..
 --   Places the new binding first, preserving whitespace.
 --   Copes with multi-line import lists
-addBindingToImportList :: T.Text -> T.Text -> T.Text
-addBindingToImportList binding importLine = case T.breakOn "(" importLine of
-    (pre, T.uncons -> Just (_, rest)) ->
-      case T.uncons (T.dropWhile isSpace rest) of
-        Just (')', _) -> T.concat [pre, "(", binding, rest]
-        _             -> T.concat [pre, "(", binding, ", ", rest]
-    _ ->
-      error
-        $  "importLine does not have the expected structure: "
+addBindingToImportList :: T.Text -> T.Text -> T.Text -> T.Text
+addBindingToImportList parent renderedBinding importLine = case T.breakOn "(" importLine of
+  (pre, T.uncons -> Just (_, rest)) ->
+    -- If the data type is in the import list wiouht the constructor, we should remove it and import it again
+    let rest' = case parent of
+          "" -> ", " <> rest
+          _ -> case T.breakOn parent rest of
+            (h, T.stripPrefix parent -> Just r) -> case T.uncons (T.dropWhile isSpace r) of
+              Just (')', _) -> ")" <> h <> r
+              Just ('(', xs) -> let imported = T.takeWhile (/= ')') xs in T.concat ["," ,imported , "), " , h , removeHeadingComma (T.tail (T.dropWhile (/= ')') r))]
+              _ -> "), " <> h <> r
+            _ -> "), " <> rest
+        binding' = (if T.null parent then id else T.init) renderedBinding
+     in removeTrailingComma $ T.concat [pre, "(", binding', rest']
```

Comment:
I find this code very hard to read. Could you try to golf it a bit to make it more readable? Some concrete suggestions:

- whitespace
- both `rest'` and `binding'` perform branching on whether `parent` is null. Make this more apparent
- introduce more names or add more comments, there's a lot of character matching that is hard to follow
- consider replacing some of the character matching with regexes if that makes things simpler.

Reply:
Thanks for your suggestions! I totally agree. It lacks of extensibility and is hard to read 😢

Reply:
@berberman when are you planning to implement this suggestion?